### ISSUE TITLE
[ENH] Avoid shadow variables

### ISSUE THREAD

**Issue by kalvdans** [cython/cython#3822]
We would like to use the -Wshadow C compiler flag to discover mistakes in our code. However, applying that flag across the board gives warnings from cython-generated code.

Here's an example

```cython
# cython: language_level = 3

def a(x):
    pass

def b():
    a(1)
```

With Cython version 0.29.17 it generates code that shadows local variables:
```
$ cython cythonshadow.pyx && gcc -Werror=shadow -I/usr/include/python3.8 cythonshadow.c  -c
cythonshadow.c: In function ‘__Pyx_PyObject_Call2Args’:
cythonshadow.c:1903:19: error: declaration of ‘args’ shadows a previous local [-Werror=shadow]
 1903 |         PyObject *args[2] = {arg1, arg2};
      |                   ^~~~
cythonshadow.c:1900:15: note: shadowed declaration is here
 1900 |     PyObject *args, *result = NULL;
      |               ^~~~
cythonshadow.c:1909:19: error: declaration of ‘args’ shadows a previous local [-Werror=shadow]
 1909 |         PyObject *args[2] = {arg1, arg2};
      |                   ^~~~
cythonshadow.c:1900:15: note: shadowed declaration is here
 1900 |     PyObject *args, *result = NULL;
      |               ^~~~
cc1: some warnings being treated as errors
```

If more people thinks this is desirable, I can prepare a pull request. Vote thumbs up! If people vote thumbs down and prefer readable generated code, let's just close the bug without action.

**Comment by da-woods** [comment]
`__Pyx_PyObject_Call2Args` has simplified considerably for Cython 3 so in the long-term this specific warning will go away. However, I think we're generally trying to remove as many C warnings as is reasonably possible, so a PR against the 0.29.x branch would be welcome.

**Comment by scoder** [comment]
> a PR against the 0.29.x branch would be welcome.

… if it shows a very low risk of breaking other people's code. If this is already fixed in the latest Cython releases, then the arguments for backporting (or creating) a fix need to be weighed off against the risk of breakage in a (very) stable release series.

Also, I'd rather like to avoid making 0.29.x diverge itself from the master branch, rather than only the other way round. That increases the risk of difficult to resolve merge conflicts and unnoticed merge accidents (such as: still works but generates dead/duplicate/useless code).

**Comment by kalvdans** [comment]
No worries, let's do it in measter.

However, I found that the code purposedly uses shadowed variables in the  __PYX_MARK_ERR_POS() macro. If this macro is expanded on a module level, it sets global variables `__pyx_filename` and `__pyx_lineno`. However, in a node level, it writes to local variables instead.

This was the first error in the test suite ran with CFLAGS="-Werror=shadow", maybe there are more places where shadowing is used as a feature.

I'll give up for now on this bug.

**Comment by scoder** [comment]
> it sets global variables __pyx_filename and __pyx_lineno. However, in a node level, it writes to local variables instead.

Yes, local variables are the right thing to use here. I think we tried removing the global variables as part of the PEP-489 / module-local state changes. Probably still lingering in an unmerged PR somewhere, but I guess it could just be done independently.

**Comment by scoder** [comment]
(In case it wasn't clear: PR welcome.)
